### DISCUSSION TITLE
Improvements to positioning on mobile for better text scanning ability

### DISCUSSION THREAD

**Discussion by StarScape** [birchill/10ten-ja-reader#1767] [Ideas]
Opening this discussion to expound on something I previously touched on with Brian in #1737.

# Motivation:

Currently, scanning through text is one of my major UX pain points on mobile. With the current behavior, scanning can be difficult because the popup positions itself unpredictably.

Say I have hovered over a word, like this one here:

<img id="Pasted-Graphic-1" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/220da3ff-6bef-4c87-b011-307dc4dd89ac" height="600px" />

If I then want to hover over some word _above_ the current line, I have a few options. If I simply go _up_ (as I might do on desktop, while still holding `Ctrl` or `Alt`), the puck moves into the popup area, but the popup does not disappear.

<img id="Pasted-Graphic-2" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/2e6f2493-9b13-4363-9e45-f73be328eb79" height="600px" />

So that’s a no-go. I can also hit the ‘X’ button to close the popup and then go up, but then I run the risk of accidentally hovering over another word on the way to the one I want. In fact, it will probably just trigger the same word again since that’s where the puck’s cursor left off.

So I’m left with awkwardly trying to weave my puck off to one side and then up the margin of the page in order to get to the word I want without accidentally triggering something.

<img id="Pasted-Graphic-3" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/7ac89646-7794-4284-8440-ae3ec2adbc34" height="600px" />

Luckily, scanning left or right isn’t an issue, so this does work. But it’s awkward and a bit frustrating. I’ve also found myself needing to do this on pages without much of a margin, which makes it harder.

You’ll notice I’ve only discussed scanning _up_, rather than down. Of course, scanning down into subsequent lines is the more common use case. You’ll mostly only find yourself scanning upwards when doubling back (which, as a not-too-advanced learner, I do quite a bit) or when skimming a page non-linearly.

Going downwards also isn’t an issue because, whenever the popup is positioned below, there seems to always be a significant margin between the word highlighted and the top of the popup:

<img id="Pasted-Graphic-5" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/12518d19-1415-456c-9447-6e645959589e" height="600px" />

<img id="Pasted-Graphic-6" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/71df38b3-de4f-4459-8668-da025092ad2e" height="600px" />

As far as I can tell, this margin only appears whenever the lookup puck is active. I assume the original intent of this was to prevent the puck itself from intersecting with the popup, but it has the added benefit of helping with scanning downwards.

But as soon as you get maybe 1/3 of the way down the page, the popup switches to positioning above the hovered word [^1]. But there’s no margin between the word and popup. And at least at the y-offset where it first starts positioning above, there’s no room for one:

<img id="Pasted-Graphic-8" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/66887f0b-10d2-486e-ac54-72d4564106a3" height="600px" />

Another factor is, since the popup appears immediately above the line I’m reading, it blocks most of the sentence the target word is in. If a user is looking up a word they don’t know, there’s a high probability they’ll want to read back over what they just read—or at least, I often do this.

Therefore, if at all possible, I think it would be desirable to keep as much of the text around the target word visible. This is particularly relevant on mobile, where dismissing the popup involves a bit more friction than on desktop.

# Potential Solutions

## Option 1: Add option to default popup position to top/bottom on mobile

**Proposed change:** Add two options to the settings menu, one for “Default popup position to top” and another for “Default popup position to bottom.” Enabling one of these would make the popup position you get by pressing `j`/`k` the default. The two options are mutually exclusive, so would likely appear as a two radio buttons hidden behind a “Change popup positioning” checkbox, greyed out if the checkbox is unchecked.

## Option 2: Add margin between hovered text and popup when popup is above

**Proposed change:** Just like there is a margin between the hovered text and popup when the popup is below the puck, add a margin when the popup is above the puck. I’m not sure how this would work when there is only just enough room to fit the popup, e.g. in the situation in the [6th screenshot](#user-content-pasted-graphic-8) above.

## Option 3: Attempt “smart positioning”

**Proposed change:** Do a sort of hybrid between the two options above. Position the popup at the top of the screen (i.e. where it would be after using the `j` shortcut) if there is enough room for the popup, plus a reasonable margin, between the hovered word and the top of the viewport. This margin should be at least large enough to see the line above, so the user could then scan up into that line, and from there into the next line above, and so on.

If there is enough for the popup plus margin at the bottom of the screen, position the popup there (i.e. where it would be after using the `k` shortcut).

If there’s no room for the popup plus margin on either above or below the hovered text (which I think could happen with a particularly large entry on a small phone screen) I think it would be best to default to positioning on top, even if that means it immediately abuts the word. Scanning downward is still a more common use case than scanning upward.

## Option 4: Popup disappears when puck intersects it

**Proposed change:** When the puck is moved up into the popup window itself, as in the [2nd screenshot](#user-content-pasted-graphic-2) above, the popup will disappear. The page text behind the popup can then be looked up as normal. Essentially, don't treat the puck like a mouse pointer, and don't have it hold the popup open.

## …And possibly others?

If you have them, please share them 🙂!

## Thoughts on potential solutions

Option 1, since there’s already code for this in the `j`/`k` shortcuts, would mostly be a matter of adding some option switches. This might also be worth doing regardless.

Option 2, if it works, would also be fairly simple, and comes with the added benefit of not adding another setting.

Option 3 is intriguing to me, because I think it has the potential to provide the best mobile UX. What I like about it is that it would leave the most text visible on either side of the hovered word, which I think would make for a better reading experience. And on mobile, I think that the benefit of displaying the popup inline is mostly negated. There’s no need for it to be nearby for mouse interaction since there is no mouse. Maybe using 10ten on a large tablet screen would change my mind, though?

Option 4 is the simplest, but I could be overlooking something that makes it non-viable. Is there any reason hovering the puck over the popup should make it remain visible? Of course it makes sense with a mouse, but I’m not sure there’s any need for it on a touch screen.

Regardless of the solution, the goal is the same: for any word on any given line, ensure that when a word is hovered, you can still move the puck cursor into the lines both _above_ and _below_.

# Conclusion

If you’ve read this far, thank you! I did not intend to spill a gallon of ink over a relatively minor UX nit. But please do share your thoughts on the above whenever you have time.

[^1]: I haven’t looked at the positioning code and and admittedly I’m not sure what the actual metric for when it switches to displaying the popup above vs below is.

**Comment by birtles** [comment]
Thanks for the detailed analysis!

For what it's worth, I found it quite hard to reproduce this issue on my Android phone but that might be a result of the particular content I was interacting with.

> I assume the original intent of this was to prevent the puck itself from intersecting with the popup, but it has the added benefit of helping with scanning downwards.

I think this actually came about because otherwise your finger would obscure the popup:

https://github.com/birchill/10ten-ja-reader/blob/42296237cab27b157f4cc1f414d3ce41345fc610/src/content/content.ts#L2287-L2297

> Therefore, if at all possible, I think it would be desirable to keep as much of the text around the target word visible. This is particularly relevant on mobile, where dismissing the popup involves a bit more friction than on desktop.

Good suggestion. I can see being able to read the prior line of text as being useful.

> ## Option 3: Attempt “smart positioning”
> 
> **Proposed change:** Do a sort of hybrid between the two options above. Position the popup at the top of the screen (i.e. where it would be after using the `j` shortcut) if there is enough room for the popup, plus a reasonable margin, between the hovered word and the top of the viewport.

I think we already do this minus the extra margin? i.e. when using the puck we prefer the `'before'` side:

https://github.com/birchill/10ten-ja-reader/blob/42296237cab27b157f4cc1f414d3ce41345fc610/src/content/popup-position.ts#L411-L431

If the popup doesn't fit on either side, we use the side that is least constrained.

> ## Option 4: Popup disappears when puck intersects it
> 
> **Proposed change:** When the puck is moved up into the popup window itself, as in the [2nd screenshot](#user-content-pasted-graphic-2) above, the popup will disappear. The page text behind the popup can then be looked up as normal. Essentially, don't treat the puck like a mouse pointer, and don't have it hold the popup open.

That certainly makes sense. I'm surprised we don't already do that!

My current take of this is:

1. We should make the popup disappear when the puck is over it (i.e. option 4). I can't think of any reason not to do that.
2. It might be useful to _also_ add more vertical margin when the popup is positioned on top so you can read the line prior to the target word. Ideally we'd only do that when there's enough space but it might complicate the positioning code too much to include that negotiation.

(Also, just in case you haven't discovered it, you can disable the puck with a single tap which will also hide the popup which might help in some cases until this issue is fixed.)

**Reply by StarScape** [reply]
> For what it's worth, I found it quite hard to reproduce this issue on my Android phone but that might be a result of the particular content I was interacting with.

It could be more of an issue for me because of screen size. I use an iPhone Mini, which is quite small. I tried on an iPad just now, and while I could still reproduce the issue, it wasn't nearly so annoying since the popup took up a much smaller segment of the screen.

> I think we already do this minus the extra margin? i.e. when using the puck we prefer the 'before' side:

Almost. What I was talking about here was always positioning the popup on either the top-left or bottom-right, never anywhere in the middle of the screen nearer the next.

So for example, not here:

<img width="400" alt="1" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/bc459841-f320-4e63-adab-9e0ae5260435">

But here instead:

<img width="400" alt="2" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/73d7dcdc-6764-4f8a-9911-314973be029d">

And not here:

<img width="400" alt="3" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/476c1c4d-a4f5-4653-a7d0-01249c85c57c">

But here instead:

<img width="400" alt="4" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/d6f893f8-1fc8-4364-b529-9ff1af896d59">

So basically, there'd be a space in the middle of the screen that is never covered by the popup:

<img width="500" alt="5" src="https://github.com/birchill/10ten-ja-reader/assets/4681820/afa2e9cc-b71a-4e1a-9c06-f13ca722c6bf">

I think this could make for a smoother reading experience on a smaller screen, but it's definitely the kind of thing I would need to test to see if it works as well in practice as it does in my head.

> My current take of this is:
> 
> 1. We should make the popup disappear when the puck is over it (i.e. option 4). I can't think of any reason not to do that.
> 2. It might be useful to also add more vertical margin when the popup is positioned on top so you can read the line prior to the target word. Ideally we'd only do that when there's enough space but it might complicate the positioning code too much to include that negotiation.

Yeah, agreed. I think implementing #1 will probably solve 90% of this annoyance. It's also a low-risk change that shouldn't involve modifying the positioning code much.

So, as long as you're cool with it, I think I'll open a PR for that soon, and then we can see from there if more vertical margin is still worth adding.

**Reply by birtles** [reply]
> Almost. What I was talking about here was always positioning the popup on either the top-left or bottom-right, never anywhere in the middle of the screen nearer the next.

Ah, I see. That makes sense. Yes, I think that would be do-able but, like you say, it's probably worth fixing the lower-hanging fruit first and then seeing if it's still necessary.

> > My current take of this is:
> > 
> > 1. We should make the popup disappear when the puck is over it (i.e. option 4). I can't think of any reason not to do that.
> > 2. It might be useful to also add more vertical margin when the popup is positioned on top so you can read the line prior to the target word. Ideally we'd only do that when there's enough space but it might complicate the positioning code too much to include that negotiation.
> 
> Yeah, agreed. I think implementing #1 will probably solve 90% of this annoyance. It's also a low-risk change that shouldn't involve modifying the positioning code much.
> 
> So, as long as you're cool with it, I think I'll open a PR for that soon, and then we can see from there if more vertical margin is still worth adding.

By all means, go for it! I guess it would involve adding code somewhere in the puck code to check if the target element is part of the popup. e.g. here: https://github.com/birchill/10ten-ja-reader/blob/42296237cab27b157f4cc1f414d3ce41345fc610/src/content/puck.ts#L443
